### ISSUE TITLE
Possible Mis-inference of Type with `source.is_ready`?

### ISSUE THREAD

**Issue by SlvrEagle23** [savonet/liquidsoap#2094]
**Describe the bug**

After updating to a relatively recent commit (`948275a22ecfd118b691613ffe6d4a87db9144e8`), I'm getting an interesting error that appears to be due to a type mis-inference, though it could possibly be something on my end.

**To Reproduce**

```
def wait_for_next_song(autodj)
    autodj_ping_attempts := !autodj_ping_attempts + 1
    delay = ref(0.5)

    if source.is_ready(!autodj) then
        log("AutoDJ is ready!")

        autodj_is_loading := false
        delay := -1.0
    elsif !autodj_ping_attempts > 200 then
        log("AutoDJ could not be initialized within the specified timeout.")

        autodj_is_loading := false
        delay := -1.0
    end

    !delay
end

dynamic = request.dynamic(id="next_song", timeout=20., retry_delay=2., autodj_next_song)
dynamic = cue_cut(id="cue_next_song", dynamic)

ref_dynamic = ref(dynamic);
thread.run.recurrent(delay=0.25, { wait_for_next_song(ref_dynamic) })
```

The error being thrown is:

```
At liquidsoap.liq, line 101, char 54-65:
   thread.run.recurrent(delay=0.25, { wait_for_next_song(ref_dynamic) })

Error 5: this value has type
  ref(source(audio=?internal('A), video=?internal('B), midi=?internal('C))) (inferred at liquidsoap.liq, line 100, char 14-26)
but it should be a subtype of the type of the value at liquidsoap.liq, line 65, char 23-29
  ref(_.{id : _})
```

For the error, the relevant lines are:

Line 100: `ref_dynamic = ref(dynamic);`

Line 65: `def wait_for_next_song(autodj)`

Basically...the only thing we're calling with that reference is `source.is_ready()`, so that should in theory be causing it to infer that it's looking for `_.{id: _}`, not a source, which it actually is looking for.

**Version details**
 - OS: Ubuntu 20.04 LTS, installed via OPAM

**Comment by smimram** [comment]
The typing system is not inferring a wrong type here, but one which is subobtimal, and it's difficult to handle this case in a generic way. Namely, since you declare
```
ref_dynamic = ref(dynamic)
```
it infers that `ref_dynamic` is a reference to the type of `dynamic`, which is a source with all the common methods (`id`, etc.). Similarly, for `autodj` it only infers that it must take a reference to a source (there is no reason why it should ask for methods such as `id`) and we have a discrepancy. The fact that it worked earlier is more due to the fact that some of the typing was overlooked. I will try to think of a "good" solution for this, but for now you can either explicitly typecast when creating the reference to force dropping the methods
```
ref_dynamic = ref((dynamic:source))
```
or have `wait_for_next_song` take the source directly and not a reference to it (it will still be "dynamic" because of the `{...}` in `thread.run.recurrent`):
```
def wait_for_next_song(autodj)
    autodj_ping_attempts := !autodj_ping_attempts + 1
    delay = ref(0.5)

    if source.is_ready(autodj) then
        log("AutoDJ is ready!")

        autodj_is_loading := false
        delay := -1.0
    elsif !autodj_ping_attempts > 200 then
        log("AutoDJ could not be initialized within the specified timeout.")

        autodj_is_loading := false
        delay := -1.0
    end

    !delay
end

dynamic = request.dynamic(id="next_song", timeout=20., retry_delay=2., autodj_next_song)
dynamic = cue_cut(id="cue_next_song", dynamic)

ref_dynamic = ref(dynamic)
thread.run.recurrent(delay=0.25, { wait_for_next_song(!ref_dynamic) })
```

**Comment by SlvrEagle23** [comment]
@smimram Thank you; the change you suggested worked without issue in my code.

It's possible all the same that this is either a regression or an undocumented change, since our previous build of 2.0.x worked with this exact same line of code. It may be worth looking into why it's triggering an error now but wasn't initially.

**Comment by smimram** [comment]
I understand that this can be ennoying, but I wouldn't really see that as a regression: this was a fortunate side effect of a bug in the typing system (which had bad consequences in other places...)
